File: backend/app/api/routes/notifications.py
from uuid import UUID

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.db.session import get_db
from app.models.user import User
from app.schemas.notification import NotificationResponse
from app.services.notifications import service as notification_service

router = APIRouter(prefix="/notifications", tags=["notifications"])


@router.get("", response_model=list[NotificationResponse])
def list_my_notifications(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> list[NotificationResponse]:
    """No role gate — every authenticated user has their own notification
    list, same reasoning as GET /approvals."""
    return notification_service.list_for_user(db, current_user)


@router.post("/{notification_id}/read", response_model=NotificationResponse)
def mark_notification_read(
    notification_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> NotificationResponse:
    return notification_service.mark_read(db, notification_id, current_user)

File: frontend/src/api/notifications.ts
import { API_BASE_URL } from './client'

export interface NotificationResponse {
  id: string
  workflow_instance_id: string | null
  type: string
  title: string
  body: string
  created_at: string
  read_at: string | null
}

interface ApiErrorBody {
  error?: { type: string; message: string }
}

/**
 * GET /notifications — no role gate, every authenticated user has their own
 * list (same pattern as GET /approvals: the filtering to "yours" happens
 * server-side, see app/api/routes/notifications.py).
 */
export async function fetchNotifications(token: string): Promise<NotificationResponse[]> {
  const response = await fetch(`${API_BASE_URL}/notifications`, {
    headers: { Authorization: `Bearer ${token}` },
  })
  if (!response.ok) {
    const body = (await response.json().catch(() => null)) as ApiErrorBody | null
    throw new Error(body?.error?.message ?? 'Failed to load notifications')
  }
  return response.json() as Promise<NotificationResponse[]>
}

export async function markNotificationRead(
  token: string,
  notificationId: string,
): Promise<NotificationResponse> {
  const response = await fetch(`${API_BASE_URL}/notifications/${notificationId}/read`, {
    method: 'POST',
    headers: { Authorization: `Bearer ${token}` },
  })
  if (!response.ok) {
    const body = (await response.json().catch(() => null)) as ApiErrorBody | null
    throw new Error(body?.error?.message ?? 'Failed to mark notification read')
  }
  return response.json() as Promise<NotificationResponse>
}
